source/request_source/request_options_list_plugin_impl.h

```diff
@@ -45,40 +44,69 @@ class RequestOptionsListRequestSource : public RequestSource {
 public:
   std::string name() const override;
 
   Envoy::ProtobufTypes::MessagePtr createEmptyConfigProto() override;
 
-  // This implementation is not thread safe. Only the first call to createRequestSourcePlugin will
-  // load the file from memory and subsequent calls just make a copy of the options_list that was
-  // already loaded. The OptionsListFromFileRequestSourceFactory will not work with multiple
-  // different files for this reason.
-  // This method will also error if the file can not be loaded correctly, e.g. the file is too large
-  // or could not be found.
+  // This implementation is not thread safe. There is a race condition because only the first call
+  // to createRequestSourcePlugin will load the file from memory and subsequent calls just make a
+  // copy of the options_list that was already loaded. The FileBasedOptionsListRequestSourceFactory
+  // will not work with multiple different files for this reason. This method will also error if the
+  // file can not be loaded correctly, e.g. the file is too large or could not be found.
   RequestSourcePtr createRequestSourcePlugin(const Envoy::Protobuf::Message& message,
                                              Envoy::Api::Api& api,
                                              Envoy::Http::RequestHeaderMapPtr header) override;
 
 private:
   Envoy::Thread::MutexBasicLockable file_lock_;
-  nighthawk::client::RequestOptionsList options_list_;
+  absl::optional<nighthawk::client::RequestOptionsList> options_list_;
 };
 
 // This factory will be activated through RequestSourceFactory in factories.h
-DECLARE_FACTORY(OptionsListFromFileRequestSourceFactory);
+DECLARE_FACTORY(FileBasedOptionsListRequestSourceFactory);
+
+// Factory that creates a OptionsListRequestSource from a InLineOptionsListRequestSourceConfig
+// proto. Registered as an Envoy plugin. Implementation of RequestSourceConfigFactory which produces
+// a RequestSource that keeps an RequestOptionsList in memory, and loads it with the RequestOptions
+// passed to it from the config. All plugins configuration are specified in the
+// request_source_plugin.proto. Usage: assume you are passed an appropriate Any type object called
+// config, an Api object called api, and a default header called header. auto& config_factory =
+//     Envoy::Config::Utility::getAndCheckFactoryByName<RequestSourcePluginConfigFactory>(
+//         "nighthawk.in-line-options-list-request-source-plugin");
+// RequestSourcePtr plugin =
+//     config_factory.createRequestSourcePlugin(config, std::move(api), std::move(header));
+
+class InLineOptionsListRequestSourceFactory : public virtual RequestSourcePluginConfigFactory {
+public:
+  std::string name() const override;
+  Envoy::ProtobufTypes::MessagePtr createEmptyConfigProto() override;
+
+  // This implementation is not thread safe. There is a race condition because only the first call
```

Comment:
I believe we have discussed this in one of the previous PRs. We shouldn't explain why it isn't thread safe as that is an implementation comment. Also the comment whether a class is or isn't thread safe belongs at class level, not method level.

Let's remove all this and instead add a single sentence at the end of the class docstring:

```cc
// This class is not thread safe.
class InLineOptionsListRequestSourceFactory : public virtual  {...}
```

Edit after looking at the current implementation - the class actually may be thread safe, because it uses a lock around this call, so that doesn't seem to be a race. However let's first discuss what we need as noted in the other comment.

Reply:
I was trying to add the clarification because of a discussion I had with Eric on this CL. I think the issue may be that I shouldn't be storing the value in the factory itself. I'm a little uncertain on this one. Essentially the problem that has resulted is that by storing the value in the factory itself, although in practice we will always get the same proto passed in from the CLI to each of the workers and thus to each call of the RequestSourcePluginConfigFactory, that technically creates a condition wherein, if multiple calls are made to the same factory with different protos, only the first one is acknowledged. In practice, as I say, this doesn't matter, but in theory this behaviour seems like some kind of race condition.

Reply:
Thank you for the clarification, sounds like all we need to do is adjust the terminology. This class actually is thread-safe. Definition of race condition is clear - a code is racy when there are multiple threads accessing the same memory and at least one of them writes. Such situation isn't possible in this code, since we are using a lock. To address this - let's update the comment at the class level and say "This class is thread-safe".

As for the interesting behavior you are describing, for now we should just document it. Can you update the documentation so that it doesn't refer to a race condition and instead mentions that this function has a specific behavior on the first call? We can choose to remove the premature optimization in a later PR.

Reply:
Understood.